llpc/util/llpcElfWriter.cpp

```diff
@@ -735,6 +737,282 @@ template <class Elf> void ElfWriter<Elf>::updateMetaNote(Context *pContext, cons
+      data += sizeof(typename Elf::Reloc);
+    }
+  }
+}
+
+// =====================================================================================================================
+// Process the reloc section
+//
+// @param reader : The ElfReader to input fragment elf bin
+// @param nonFragmentPsIsaOffset : The offset of PS ISA in nonFragment elf bin
+// @param fragmentPsIsaOffset : The offset of PS ISA in input fragment elf bin
+template <class Elf>
+void ElfWriter<Elf>::processRelocSection(const ElfReader<Elf> &reader, size_t nonFragmentPsIsaOffset,
+                                         size_t fragmentPsIsaOffset) {
+
+  ElfSectionBuffer<Elf64::SectionHeader> *fragmentRelocSection = nullptr;
+  const ElfSectionBuffer<Elf64::SectionHeader> *nonFragmentRelocSection = nullptr;
+
+  auto fragmentRelocSecIndex = reader.GetSectionIndex(RelocName);
+  auto nonFragmentRelocSecIndex = GetSectionIndex(RelocName);
+
+  reader.getSectionDataBySectionIndex(fragmentRelocSecIndex, &fragmentRelocSection);
+  getSectionDataBySectionIndex(nonFragmentRelocSecIndex, &nonFragmentRelocSection);
+
+  bool isFragmentRelocUsedInPs = false;
+
+  // The start pos of the relocs which used by PS.
+  size_t fragmentRelocPsStartPos = 0;
+
+  if (fragmentRelocSection) {
+    size_t fragmentRelocCounts = numRelocs(fragmentRelocSection);
+    fragmentRelocPsStartPos = getRelocPsStartPos(fragmentRelocSection, fragmentPsIsaOffset);
+    isFragmentRelocUsedInPs = (fragmentRelocPsStartPos < fragmentRelocCounts);
+  }
+
+  if (!nonFragmentRelocSection && isFragmentRelocUsedInPs) {
+    // The input fragment elf bin has relocs used in PS, but nonFragment elf bin does not have reloc section,
+    // then a reloc section need to be created for nonFragment elf bin.
+    SectionBuffer newRelocSection = {};
+
+    createNewSection(RelocName, fragmentRelocSection, &newRelocSection);
+
+    m_sections.push_back(newRelocSection);
+    // Update the reloc section index.
+    m_relocSecIdx = m_sections.size() - 1;
+    m_map[RelocName] = m_relocSecIdx;
+    nonFragmentRelocSecIndex = m_relocSecIdx;
+    getSectionDataBySectionIndex(nonFragmentRelocSecIndex, &nonFragmentRelocSection);
+  }
+
+  // Merge reloc section.
+  if (nonFragmentRelocSection) {
+
+    size_t nonFragmentRelocPsStartPos = getRelocPsStartPos(nonFragmentRelocSection, nonFragmentPsIsaOffset);
+
+    // If the input fragment elf bin does not have relocs used by PS,
+    // no need to merge any reloc from the input fragment elf bin.
+    ElfSectionBuffer<Elf64::SectionHeader> *tmpFragmentRelocSection =
+        isFragmentRelocUsedInPs ? fragmentRelocSection : nullptr;
+
+    int64_t diffOfPsOffset = nonFragmentPsIsaOffset - fragmentPsIsaOffset;
+
+    ElfSectionBuffer<Elf64::SectionHeader> newSection = {};
+
+    mergeRelocSection(reader, nonFragmentRelocSection, nonFragmentRelocPsStartPos, tmpFragmentRelocSection,
+                      fragmentRelocPsStartPos, diffOfPsOffset, &newSection);
```

Comment:
Can you make `mergeSection` return `SectionHeader` instead of using an output parameter? (Same like `createNewSection`)

Reply:
Sure.